torch/utils/data/graph.py

```diff
@@ -52,11 +52,18 @@ def reduce_hook(obj):
 
 
 def traverse(datapipe, only_datapipe=False):
+    cache: Set[IterDataPipe] = set()
+    return _traverse_helper(datapipe, only_datapipe, cache)
+
+
+# Add cache here to prevent infinite recursion on DataPipe
+def _traverse_helper(datapipe, only_datapipe=False, cache=None):
     if not isinstance(datapipe, IterDataPipe):
         raise RuntimeError("Expected `IterDataPipe`, but {} is found".format(type(datapipe)))
 
-    items = list_connected_datapipes(datapipe, only_datapipe)
+    cache.add(datapipe)
+    items = list_connected_datapipes(datapipe, only_datapipe, cache)
     d: Dict[IterDataPipe, Any] = {datapipe: {}}
     for item in items:
-        d[datapipe].update(traverse(item, only_datapipe))
+        d[datapipe].update(_traverse_helper(item, only_datapipe, cache.copy()))
```

Comment:
Why do we need the copy here?

Reply:
Yeah. We are checking if `IterDataPipe` exists in local path rather than global path. The following structure should support a single `IterDataPipe` being traversed multiple times.
```
dp0 -> dp1 -> dp2 -> dp4
         \
          v
          dp3 -> dp5
```
`dp0` and `dp1` should be traversed twice IMO.